File: Assets/AltTester/Runtime/AltDriver/AltDriverUnity.cs
/*
    Copyright(C) 2025 Altom Consulting

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/>.
*/

using System;
using System.Collections.Generic;
using AltTester.AltTesterUnitySDK.Driver.Commands;

namespace AltTester.AltTesterUnitySDK.Driver
{
    public class AltDriverUnity : AltDriverBase
    {
        public AltDriverUnity(string host = "127.0.0.1", int port = 13000, string appName = "__default__", bool enableLogging = false, int connectTimeout = 60, string platform = "unknown", string platformVersion = "unknown", string deviceInstanceId = "unknown", string appId = "unknown", string driverType = "SDK")
            : base(host, port, appName, enableLogging, connectTimeout, platform, platformVersion, deviceInstanceId, appId, driverType)
        {
        }

        public new void ResetInput()
        {
            base.ResetInput();
        }
        public new void LoadScene(string scene, bool loadSingle = true)
        {
            base.LoadScene(scene, loadSingle);
        }

        public new void UnloadScene(string scene)
        {
            base.UnloadScene(scene);
        }

        public new List<string> GetAllLoadedScenes()
        {
            return base.GetAllLoadedScenes();
        }

        public new string GetCurrentScene()
        {
            return base.GetCurrentScene();
        }

        public new void WaitForCurrentSceneToBe(string sceneName, double timeout = 10, double interval = 1)
        {
            base.WaitForCurrentSceneToBe(sceneName, timeout, interval);
        }

        public new void SetTimeScale(float timeScale)
        {
           base.SetTimeScale(timeScale);
        }

        public new float GetTimeScale()
        {
            return base.GetTimeScale();
        }


        public new void DeletePlayerPref()
        {
            base.DeletePlayerPref();
        }

        public new void DeleteKeyPlayerPref(string keyName)
        {
            base.DeleteKeyPlayerPref(keyName);
        }

        public new void SetKeyPlayerPref(string keyName, int valueName)
        {
            base.SetKeyPlayerPref(keyName, valueName);
        }

        public new void SetKeyPlayerPref(string keyName, float valueName)
        {
            base.SetKeyPlayerPref(keyName, valueName);
        }

        public new void SetKeyPlayerPref(string keyName, string valueName)
        {
            base.SetKeyPlayerPref(keyName, valueName);
        }

        public new int GetIntKeyPlayerPref(string keyName)
        {
            return base.GetIntKeyPlayerPref(keyName);
        }

        public new float GetFloatKeyPlayerPref(string keyName)
        {
            return base.GetFloatKeyPlayerPref(keyName);
        }

        public new string GetStringKeyPlayerPref(string keyName)
        {
           return base.GetStringKeyPlayerPref(keyName);
        }

        public new List<AltObject> FindObjects(AltBy altBy, AltBy cameraAltBy = null, bool enabled = true)
        {
            return base.FindObjects(altBy, cameraAltBy, enabled);
        }

        public new List<AltObject> FindObjectsWhichContain(AltBy altBy, AltBy cameraAltBy = null, bool enabled = true)
        {
            return base.FindObjectsWhichContain(altBy, cameraAltBy, enabled);
        }

        public new AltObject FindObject(AltBy altBy, AltBy cameraAltBy = null, bool enabled = true)
        {
            return base.FindObject(altBy, cameraAltBy, enabled);
        }

        public new AltObject FindObjectWhichContains(AltBy altBy, AltBy cameraAltBy = null, bool enabled = true)
        {
            return base.FindObjectWhichContains(altBy, cameraAltBy, enabled);
        }

        public new T CallStaticMethod<T>(string typeName, string methodName, string assemblyName,
                    object[] parameters, string[] typeOfParameters = null)
        {
            return base.CallStaticMethod<T>(typeName, methodName, assemblyName, parameters, typeOfParameters);
        }

        public new T GetStaticProperty<T>(string componentName, string propertyName, string assemblyName, int maxDepth = 2)
        {
            return base.GetStaticProperty<T>(componentName, propertyName, assemblyName, maxDepth);
        }

        public new void SetStaticProperty(string componentName, string propertyName, string assemblyName, object updatedProperty)
        {
            base.SetStaticProperty(componentName, propertyName, assemblyName, updatedProperty);
        }

        /// <summary>
        /// Simulates a swipe action between two points.
        /// </summary>
        /// <param name="start">Coordinates of the screen where the swipe begins</param>
        /// <param name="end">Coordinates of the screen where the swipe ends</param>
        /// <param name="duration">The time measured in seconds to move the mouse from start to end location. Defaults to <c>0.1</c>.</param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void Swipe(AltVector2 start, AltVector2 end, float duration = 0.1f, bool wait = true)
        {
            base.Swipe(start, end, duration, wait);
        }

        /// <summary>
        /// Simulates a multipoint swipe action.
        /// </summary>
        /// <param name="positions">A list of positions on the screen where the swipe be made.</param>
        /// <param name="duration">The time measured in seconds to swipe from first position to the last position. Defaults to <code>0.1</code>.</param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void MultipointSwipe(AltVector2[] positions, float duration = 0.1f, bool wait = true)
        {
           base.MultipointSwipe(positions, duration, wait);
        }

        /// <summary>
        /// Simulates holding left click button down for a specified amount of time at given coordinates.
        /// </summary>
        /// <param name="coordinates">The coordinates where the button is held down.</param>
        /// <param name="duration">The time measured in seconds to keep the button down.</param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void HoldButton(AltVector2 coordinates, float duration, bool wait = true)
        {
            base.HoldButton(coordinates, duration, wait);
        }

        /// <summary>
        /// Simulates key press action in your app.
        /// </summary>
        /// <param name="keyCode">The key code of the key simulated to be pressed.</param>
        /// <param name="power" >A value between [-1,1] used for joysticks to indicate how hard the button was pressed. Defaults to <c>1</c>.</param>
        /// <param name="duration">The time measured in seconds from the key press to the key release. Defaults to <c>0.1</c></param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void PressKey(AltKeyCode keyCode, float power = 1, float duration = 0.1f, bool wait = true)
        {
            base.PressKey(keyCode, power, duration, wait);
        }

        /// <summary>
        /// Simulates multiple keys pressed action in your app.
        /// </summary>
        /// <param name="keyCodes">The list of key codes of the keys simulated to be pressed.</param>
        /// <param name="power" >A value between [-1,1] used for joysticks to indicate how hard the button was pressed. Defaults to <c>1</c>.</param>
        /// <param name="duration">The time measured in seconds from the key press to the key release. Defaults to <c>0.1</c></param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void PressKeys(AltKeyCode[] keyCodes, float power = 1, float duration = 0.1f, bool wait = true)
        {
            base.PressKeys(keyCodes, power, duration, wait);
        }

        public new void KeyDown(AltKeyCode keyCode, float power = 1)
        {
            base.KeyDown(keyCode, power);
        }

        /// <summary>
        /// Simulates multiple keys down action in your app.
        /// </summary>
        /// <param name="keyCodes">The key codes of the keys simulated to be down.</param>
        /// <param name="power" >A value between [-1,1] used for joysticks to indicate how hard the button was pressed. Defaults to <c>1</c>.</param>
        public new void KeysDown(AltKeyCode[] keyCodes, float power = 1)
        {
            base.KeysDown(keyCodes, power);
        }


        /// <summary>
        /// Simulate mouse movement in your app.
        /// </summary>
        /// <param name="coordinates">The screen coordinates</param>
        /// <param name="duration">The time measured in seconds to move the mouse from the current mouse position to the set coordinates. Defaults to <c>0.1f</c></param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void MoveMouse(AltVector2 coordinates, float duration = 0.1f, bool wait = true)
        {
            base.MoveMouse(coordinates, duration, wait);
        }

        /// <summary>
        /// Simulate scroll action in your app.
        /// </summary>
        /// <param name="speed">Set how fast to scroll. Positive values will scroll up and negative values will scroll down. Defaults to <code> 1 </code></param>
        /// <param name="duration">The duration of the scroll in seconds. Defaults to <code> 0.1 </code></param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void Scroll(float speed = 1, float duration = 0.1f, bool wait = true)
        {
            base.Scroll(speed, duration, wait);
        }

        /// <summary>
        /// Simulate scroll action in your app.
        /// </summary>
        /// <param name="scrollValue">Set how fast to scroll. X is horizontal and Y is vertical. Defaults to <code> 1 </code></param>
        /// <param name="duration">The duration of the scroll in seconds. Defaults to <code> 0.1 </code></param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void Scroll(AltVector2 scrollValue, float duration = 0.1f, bool wait = true)
        {
            base.Scroll(scrollValue, duration, wait);
        }

        /// <summary>
        /// Tap at screen coordinates
        /// </summary>
        /// <param name="coordinates">The screen coordinates</param>
        /// <param name="count">Number of taps</param>
        /// <param name="interval">Interval between taps in seconds</param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void Tap(AltVector2 coordinates, int count = 1, float interval = 0.1f, bool wait = true)
        {
            base.Tap(coordinates, count, interval, wait);
        }

        /// <summary>
        /// Simulates device rotation action in your app.
        /// </summary>
        /// <param name="acceleration">The linear acceleration of a device.</param>
        /// <param name="duration">How long the rotation will take in seconds. Defaults to <code>0.1<code>.</param>
        /// <param name="wait">If set wait for command to finish. Defaults to <c>True</c>.</param>
        public new void Tilt(AltVector3 acceleration, float duration = 0.1f, bool wait = true)
        {
            base.Tilt(acceleration, duration, wait);
        }

        public new List<AltObject> GetAllElements(AltBy cameraAltBy = null, bool enabled = true)
        {
            return base.GetAllElements(cameraAltBy, enabled);
        }

        public new List<AltObjectLight> GetAllElementsLight(AltBy cameraAltBy = null, bool enabled = true)
        {
            return base.GetAllElementsLight(cameraAltBy, enabled);
        }

        public new AltObject WaitForObject(AltBy altBy, AltBy cameraAltBy = null, bool enabled = true, double timeout = 20, double interval = 0.5)
        {
            return base.WaitForObject(altBy, cameraAltBy, enabled, timeout, interval);
        }

        public new void WaitForObjectNotBePresent(AltBy altBy, AltBy cameraAltBy = null, bool enabled = true, double timeout = 20, double interval = 0.5)
        {
            base.WaitForObjectNotBePresent(altBy, cameraAltBy, enabled, timeout, interval);
        }

        public new AltObject WaitForObjectWhichContains(AltBy altBy, AltBy cameraAltBy = null, bool enabled = true, double timeout = 20, double interval = 0.5)
        {
            return base.WaitForObjectWhichContains(altBy, cameraAltBy, enabled, timeout, interval);
        }

        public new List<string> GetAllScenes()
        {
            return base.GetAllScenes();
        }

        public new List<AltObject> GetAllCameras()
        {
            return base.GetAllCameras();
        }

        public new List<AltObject> GetAllActiveCameras()
        {
            return base.GetAllActiveCameras();
        }
        public new AltVector2 GetApplicationScreenSize()
        {
            return base.GetApplicationScreenSize();
        }


        public new int BeginTouch(AltVector2 screenPosition)
        {
           return base.BeginTouch(screenPosition);
        }

        public new void MoveTouch(int fingerId, AltVector2 screenPosition)
        {
            base.MoveTouch(fingerId, screenPosition);
        }

        public new void EndTouch(int fingerId)
        {
            base.EndTouch(fingerId);
        }

        public new AltObject FindObjectAtCoordinates(AltVector2 screenPosition)
        {
            return base.FindObjectAtCoordinates(screenPosition);
        }
    }
}

File: Bindings~/python/alttester/altdriver_unity.py
"""
    Copyright(C) 2024 Altom Consulting

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/>.
"""

import warnings
from alttester.altobject import AltObject
from alttester.altdriver import AltDriver
from alttester.altby import AltBy
from alttester.altdriver_base import AltDriverBase

warnings.filterwarnings("default", category=DeprecationWarning, module=__name__)


class AltDriverUnity(AltDriverBase):
    """The driver object will help interacting with all the application objects, their properties and methods.

    When you instantiate an ``AltDriverUnity`` object in your tests, you can use it to “drive” your application like
    one of your users would, by interacting with all the application objects, their properties and methods.
    An ``AltDriverUnity`` instance will connect to the AltTester® Server.

    Args:
        host (:obj:`str`, optional): The host to connect to.
        port (:obj:`int`, optional): The port to connect to.
        app_name (:obj:`str`, optional): The application name. Defaults to ``__default__``.
        enable_logging (:obj:`bool`, optional): If set to ``True`` will turn on logging, by default logging is disabled.
        timeout (:obj:`int`, :obj:`float`, optional): The timeout duration for establishing a connection, in seconds.
            If set to ``None``, the connection attempt will wait indefinitely. The default value is ``60`` seconds.

    """

    def __init__(
        self,
        host="127.0.0.1",
        port=13000,
        app_name="__default__",
        enable_logging=False,
        timeout=60,
        platform="unknown",
        platform_version="unknown",
        device_instance_id="unknown",
        app_id="unknown",
    ):
        super().__init__(
            host,
            port,
            app_name,
            enable_logging,
            timeout,
            platform,
            platform_version,
            device_instance_id,
            app_id,
        )

    def find_object(
        self, altby: AltBy, camera_altby: AltBy = AltBy.name(""), enabled: bool = True
    ) -> AltObject:
        # """Finds an object in the scene.

        # Args:
        #     altby (:obj:`AltBy`): The search criteria.
        #     camera_altby (:obj:`AltBy`, optional): The camera search criteria.
        #     enabled (:obj:`bool`, optional): If set to ``True``, the search will only return enabled objects.

        # Returns:
        #     :obj:`AltObject`: The object found.

        # Raises:
        #     :obj:`Exception`: If the object is not found.

        # """
        return self._AltDriverBase__find_object(altby, camera_altby, enabled)

    def find_objects(
        self, altby: AltBy, camera_altby: AltBy = AltBy.name(""), enabled: bool = True
    ) -> list:
        """Finds all objects in the scene.

        Args:
            altby (:obj:`AltBy`): The search criteria.
            camera_altby (:obj:`AltBy`, optional): The camera search criteria.
            enabled (:obj:`bool`, optional): If set to ``True``, the search will only return enabled objects.

        Returns:
            :obj:`list`: The list of objects found.

        Raises:
            :obj:`Exception`: If the object is not found.

        """
        return self._AltDriverBase__find_objects(altby, camera_altby, enabled)

    def find_object_which_contains(
        self, altby: AltBy, camera_altby: AltBy = AltBy.name(""), enabled: bool = True
    ) -> AltObject:
        """Finds an object in the scene which contains the search criteria.

        Args:
            altby (:obj:`AltBy`): The search criteria.
            camera_altby (:obj:`AltBy`, optional): The camera search criteria.
            enabled (:obj:`bool`, optional): If set to ``True``, the search will only return enabled objects.

        Returns:
            :obj:`AltObject`: The object found.

        Raises:
            :obj:`Exception`: If the object is not found.

        """
        return self._AltDriverBase__find_object_which_contains(altby, camera_altby, enabled)

    def find_objects_which_contain(
        self, altby: AltBy, camera_altby: AltBy = AltBy.name(""), enabled: bool = True
    ) -> list:
        """Finds all objects in the scene which contain the search criteria.

        Args:
            altby (:obj:`AltBy`): The search criteria.
            camera_altby (:obj:`AltBy`, optional): The camera search criteria.
            enabled (:obj:`bool`, optional): If set to ``True``, the search will only return enabled objects.

        Returns:
            :obj:`list`: The list of objects found.

        Raises:
            :obj:`Exception`: If the object is not found.

        """
        return self._AltDriverBase__find_objects_which_contain(altby, camera_altby, enabled)

    def wait_for_object(
        self,
        altby: AltBy,
        camera_altby: AltBy = AltBy.name(""),
        timeout: int = 20,
        interval: int = 0.5,
        enabled: bool = True
    ) -> AltObject:
        """Waits for an object to be found in the scene.

        Args:
            altby (:obj:`AltBy`): The search criteria.
            camera_altby (:obj:`AltBy`, optional): The camera search criteria.
            enabled (:obj:`bool`, optional): If set to ``True``, the search will only return enabled objects.
            timeout (:obj:`int`, optional): The timeout duration for the search, in seconds.

        Returns:
            :obj:`AltObject`: The object found.

        Raises:
            :obj:`Exception`: If the object is not found.

        """
        return self._AltDriverBase__wait_for_object(altby, camera_altby, timeout, interval, enabled)

    def wait_for_object_which_contains(
        self,
        altby: AltBy,
        camera_altby: AltBy = AltBy.name(""),
        timeout: int = 20,
        interval: int = 0.5,
        enabled: bool = True
    ) -> AltObject:
        """Waits for an object to be found in the scene which contains the search criteria.

        Args:
            altby (:obj:`AltBy`): The search criteria.
            camera_altby (:obj:`AltBy`, optional): The camera search criteria.
            enabled (:obj:`bool`, optional): If set to ``True``, the search will only return enabled objects.
            timeout (:obj:`int`, optional): The timeout duration for the search, in seconds.

        Returns:
            :obj:`AltObject`: The object found.

        Raises:
            :obj:`Exception`: If the object is not found.

        """
        return self._AltDriverBase__wait_for_object_which_contains(altby, camera_altby, enabled, timeout, interval)

    def wait_for_object_to_not_be_present(
        self,
        altby: AltBy,
        camera_altby: AltBy = AltBy.name(""),
        timeout: int = 20,
        interval: int = 0.5,
        enabled: bool = True,
    ) -> None:
        """Waits for an object not to be present in the scene.

        Args:
            altby (:obj:`AltBy`): The search criteria.
            camera_altby (:obj:`AltBy`, optional): The camera search criteria.
            timeout (:obj:`int`, optional): The timeout duration for the search, in seconds.

        Raises:
            :obj:`Exception`: If the object is found.

        """
        return self._AltDriverBase__wait_for_object_to_not_be_present(altby, camera_altby, timeout, interval, enabled)

    def get_all_elements(self, camera_altby: AltBy = AltBy.name(""), enabled: bool = True) -> list:
        """Returns all the elements in the scene.

        Args:
            camera_altby (:obj:`AltBy`, optional): The camera search criteria.

        Returns:
            :obj:`list`: The list of objects found.

        """
        return self._AltDriverBase__get_all_elements(camera_altby, enabled)